### PR TITLE
Add OIDC token grace period, better support for namespace-qualified claim paths and redirects

### LINKED ISSUE #6836: Provide an OIDC token expiry grace period property
**Description**
Vertx Oauth2ClientOptions has a `JwtOptions` property which can be used to configure a `leeway` property which is an expiry grace period.

### LINKED ISSUE #6842: OIDC and smallrye-jwt are unable to get groups from custom claims like "https://example.com/roles"
**Describe the bug**
Auth0 only allows adding custom claims to accessToken with a namespaced format, such as "https://example.com/roles": ["user", "admin"]
As documented at  https://auth0.com/docs/api-auth/tutorials/adoption/scope-custom-claims#custom-claims 

Quarkus-oidc and smallrye-token fail to get the value of this custom claim.
As workaround I added these roles in the scope attribute.

**Expected behavior**
OIDC and smallrye-jwt are able to obtain groups from custom claim  such as "https://example.com/roles"

**Actual behavior**
OIDC and smallrye-jwt are unable to obtain the claims value and output the following log:
```
2020-01-28 21:01:59,623 WARN  [io.sma.jwt.aut.pri.DefaultJWTTokenParser] (vert.x-eventloop-thread-3) No claim exists at the path https://example.com/roles at segment https:
2020-01-28 21:01:59,623 WARN  [io.sma.jwt.aut.pri.DefaultJWTTokenParser] (vert.x-eventloop-thread-3) Claim value at the path https://example.com/roles is neither an array of strings nor string
```

**Configuration**
```properties
smallrye.jwt.path.groups=https://example.com/roles

```
**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:  Linux jom-mnt19 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
 - Output of `java -version`:  openjdk version "1.8.0_232" - OpenJDK 64-Bit Server VM (build 25.232-b09, mixed mode)
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.1.1

### LINKED ISSUE #6853: OIDC relative redirect-path property should not be tied in to the current relative request path
**Description**
`redirect-path` works only if it is confirmed to be part of the request URI, example, given `https://host:8080/admin/1` and a redirect path `/admin`, the code flow `redirect_uri` parameter will become `https://host:8080/admin`. It does not work if it is set to some sibling path like `callback` in which case `redirect_uri` should become `https://host:8080/callback`.

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #6853.
(`redirect-path` support on the master is very limited: not possible to redirect to any other path but the one which the current request path starts from)
Fixes #6842.
(namespace qualified claim paths)
Fixes #6836.
(Token expiration grace period)

Also adds an ability to set the extra authentication request parameters as asked by @emmanuelBayle

### CODE REVIEW

**Review by pedroigor — CHANGES_REQUESTED** [review]
Nice work @sberyozkin. What about tests for redirectUri and extra params ?

**Comment by sberyozkin** [comment]
@pedroigor Hey, thanks, `redirect_uri` is already being tested with the `oidc_code_flow`, see `application.properties`. With these changes it is actually all the `CodeFlowTest` tests now that are affected. Not really sure to assert that it has been effective, may be check the `q_session` cookie as it saves the original path by default, I'll have a look.
Yes, good idea re the extra parameters. Speaking of them, do you know which of the extra parameters which are linked to in the PR Keycloak will react to (such that for example I can check the token content has been affected) ?
thanks

**Comment by pedroigor** [comment]
@sberyozkin for the extra params, we could force a `prompt=login` and see if the login page is shown even though the user is authenticated.

**Comment by sberyozkin** [comment]
@pedroigor That would require some html unit checks, I'd rather not touch it, but I'll experiment, thanks

**Comment by sberyozkin** [comment]
@pedroigor Hi Pedro, I've updated `CodeFlowTest` to check `Location` (only for 1 test but I reckon it is sufficient as I'd like to avoid stressing htmlunit too much :-) ) and the `q_auth` cookie for all the tests - which checks what the actual `redirect_uri` parameter was and also what the saved path in the cookie was, and also an extra `max-age` param is checked.
How does it look now ?

**Comment by sberyozkin** [comment]
@pedroigor @stuartwdouglas I've done another test improvement (the state parameter is now properly checked) and added a minor optimization around the pattern (split only if a `/` character is found)

**Comment by sberyozkin** [comment]
Would be good to get it to `1.3.0.Alpha1` as a few users are waiting. I've found an issue though related to the multi-tenancy and the code flow, but it is a bigger and orthogonal issue

**Review by pedroigor — APPROVED** [review]

**Comment by sberyozkin** [comment]
@pedroigor Hey Pedro, thanks for the approval, in meantime I've also added one more test, which tests the case of the path not being restored, using the multitenancy feature, which is really cool, but I'll need to open an issue to follow up with a few improvements which are not related to this actual issue.
Thanks for asking me to test better, it was a really lame initial PR, sorry :-). Now I'm happy. Stuart might not be available so I'll merge once the tests pass, your approval is sufficient.

Thanks

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java
Comment (gsmet) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java]: @sberyozkin I let you change that one to use `debugf`. Ping me when it's done. Thanks!

**Review by gsmet — CHANGES_REQUESTED** [review]

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java]: @gsmet Hi Guillaume, it is done, there were 2 more instances of `debug` in some other parts of this class. Thanks for spotting it.

**Review by gastaldi — APPROVED** [review]

**Review by gsmet — APPROVED** [review]

**Comment by sberyozkin** [comment]
@gastaldi @gsmet Thanks :-)

**Pull request merged**
